### PR TITLE
Use stateful dataloader to checkpoint data iteration order and token buffer

### PR DESCRIPTION
**Pull request by gokulavasan**
Summary: 

Use the stateful_dataloader from torchdata (https://github.com/pytorch/data/tree/main/torchdata/stateful_dataloader) for storing the token buffer and iteration data order. It requires a dependency on the nightly build of torchdata >= 20240426.

Also make sure the dataloader state has a different key per rank.

Test Plan:

Tested locally by first running 30 steps (checkpointing every 5 steps) and capturing all the loss values. Then deleting the last 3 checkpoints and then re-run the training and the loss values from step 16-30 match with what we had earlier in the first run. Note that this requires changes in the train.py to enable a deterministic run.

Reviewers: @tianyu-l 

Subscribers: @andrewkho 

Tasks:

Tags:

### CODE REVIEW

File: torchtitan/checkpoint.py
Comment (andrewkho) [inline torchtitan/checkpoint.py:25]: Why do we need this too?

File: torchtitan/datasets/hf_datasets.py
Comment (andrewkho) [inline torchtitan/datasets/hf_datasets.py]: Does this happen at end-of-epoch? In this case should we just restart the iterator instead of returning empty iterator?

File: torchtitan/checkpoint.py
Comment (gokulavasan) [reply torchtitan/checkpoint.py:25]: @andrewkho For typing. I can switch to StatefulDataLoader if we know that is going to be the one used in all scenarios

File: torchtitan/datasets/hf_datasets.py
Comment (gokulavasan) [reply torchtitan/datasets/hf_datasets.py]: @andrewkho Yes, this is for end-of-epoch. I structured this method to take care of creating a skipped iterator. The caller takes care of restarting if self.infinite is set to true.

**Comment by gokulavasan** [comment]
Looks like adding the index url (for torchdata) is causing other dependencies to not get installed. Will figure out how to fix this

File: torchtitan/datasets/hf_datasets.py
Comment (fegin) [inline torchtitan/datasets/hf_datasets.py:160]: Is the `_all_tokens` a DTensor? If not, only rank0's dataloader data will be saved.

Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:160]: Same as `_sample_idx`.

Comment (tianyu-l) [reply torchtitan/datasets/hf_datasets.py:160]: This is the state dict of the dataset, and it will be wrapped by a StatefulDataloader which also has state dict calling into this.  In `checkpoint.py`, the StatefulDataloader is further wrapped by another stateful wrapper with different keys for different ranks. Would it be OK this way? @fegin

Comment (tianyu-l) [reply torchtitan/datasets/hf_datasets.py:160]: We think the final solution should use DTensor, and this PR is more like the first version.

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py:143]: Curious why we are not using HF's `skip()` for both iterable and map-style.

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py]:
```suggestion
            # Skip samples for the first iteration to resume from checkpoint properly
```

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py]: nit: There are many variables in this class. Shall we group `self._sample_idx`, `self._all_tokens`, and `self._first_iter` together at the bottom of `__init__`, and comment something like `# variables needed for checkpointing`.

File: torchtitan/datasets/hf_datasets.py
Comment (gokulavasan) [reply torchtitan/datasets/hf_datasets.py:160]: @fegin For now, the state_dict returned by the dataloader is keyed by the rankid. That is {0: state_dict}. Would that ensure the state of all the ranks is stored? Whether we want to use DTensor is something we are looking into (as resharding dataloader state is a challenge anyway so the benefit of storing it as DTensor is a bit limited).

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py:155]: Since we expect this class to have `state_dict` and `load_state_dict` methods, it might be better to let the class inherit a Stateful interface:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/stateful.py

File: requirements.txt
Comment (tianyu-l) [inline requirements.txt]: What is the earliest nightly that is good to test? Apr 25?

We depend on nightly pytorch in CI, so this can be removed when we land this PR.
https://github.com/pytorch/torchtitan/blob/main/.github/workflows/unit_test_4gpu.yaml#L36

**Review by tianyu-l — COMMENTED** [review]
First pass looks awesome already. Let me see if I can run some large scale experiments. What is the earliest nightly that's OK to run with this PR?

File: torchtitan/datasets/hf_datasets.py
Comment (fegin) [reply torchtitan/datasets/hf_datasets.py:160]: As long as the key is unique across ranks, using DTensor is not required.

File: requirements.txt
Comment (gokulavasan) [reply requirements.txt]: @tianyu-l Need torchdata build from Apr 26 (20240426) at least

Comment (tianyu-l) [reply requirements.txt]: I think we still need the `torchdata` dependency in requirements.txt.

For HF datasets it seems from local test that we need `datasets >= 2.19.0` to use the `skip()` API

Comment (fegin) [reply requirements.txt]: confirm that I need to reinstall datasets to run with the PR

**Comment by fegin** [comment]
@gokulavasan, @tianyu-l https://github.com/pytorch/pytorch/pull/125335 and https://github.com/pytorch/pytorch/pull/125334 should unblock this PR.

File: torchtitan/checkpoint.py
Comment (rlrs) [inline torchtitan/checkpoint.py]: This seems to break because `rank_id` gets cast to a string at some point. Should maybe cast it to a string manually in both functions?

**Comment by rlrs** [comment]
I've been testing this out, and ran into an issue with resuming from a checkpoint. I suspect it's because of how `StatefulDataLoader` handles the state dict: https://github.com/pytorch/data/blob/11e16da61d7f5f587627c75e99ea664efef3e0f8/torchdata/stateful_dataloader/stateful_dataloader.py#L249

That is, a freshly initialized `StatefulDataLoader` does not have a state dict to load into? I'm not very familiar with how `DCP` works, so please correct me if it's wrong.

**Comment by gokulavasan** [comment]
@rlrs Would it be possible to test it after my latest commit ([b9b045d](https://github.com/pytorch/torchtitan/pull/279/commits/b9b045d32933c2824ae6f667e944a51c3255a2d1)? I missed addin that part.

**Comment by rlrs** [comment]
> @rlrs Would it be possible to test it after my latest commit ([b9b045d](https://github.com/pytorch/torchtitan/pull/279/commits/b9b045d32933c2824ae6f667e944a51c3255a2d1))? I missed adding that part.

I had already added that in my version. I can't get it to load the state_dict, unless I first call `iter(dataloader)` so that `dataloader._iterator is not None`. 

If I call `iter` before `DCP.load`, and then set `self._first_iter = True` in `HuggingFaceDataset.load_state_dict`, everything seems to work!

File: requirements.txt
Comment (tianyu-l) [inline requirements.txt]: I think we need to put the `torchdata` dependency here, maybe try
```suggestion
torch >= 2.2.0.dev
--find-links https://download.pytorch.org/whl/nightly/cpu/
torchdata >= 0.7.1.dev20240426+cpu
```
If works, we can remove the dependency in unit test workflow.

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py]: Let's put all of them into the helper function to make it cleaner, e.g. with name `_get_data_iter` or `_get_maybe_skipped_data_iter`. Also this function doesn't need input, because `self._sample_idx` and `self._first_iter` are both class variables.

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py]: IIRC, `len(self._data)` is only available for map-style?

**Review by tianyu-l — COMMENTED** [review]
Looks great! Had some minor comments.
For next steps, we can discuss if we should add a unit test to guard the correctness of checkpointable data loading, and the plan to migrate to DTensor-based checkpointing.

File: requirements.txt
Comment (gokulavasan) [reply requirements.txt]: I attempted thus but it fails to find the nightly, instead installs the latest released version. We do plan to do a torchdata release soon (which will contain the latest StatefulDataLoader) and once that happens, we can remove the explicit pip installs in unit test workflows and main README.

**Comment by gokulavasan** [comment]
@tianyu-l Addressed PR comments (thank you!), added unit test, and made changes to the github workflows to allow running those unit tests. Let me know if the changes look okay. Regarding the move to DTensor, I think this requires analysis of what the benefits are (especially for storing unstructured state dict of dataloader).

If it is purely to reduce the replication of state across tensor/pipeline parallel groups, I think we can store the dataloader state just for the dp worker ranks (by using key as the dp_rank_id) and load it back instead of storing it for all global ranks. For now, with just the text tokens, this might not even be necessary as the state is not that big. Let me know how you would like to proceed.

**Comment by gokulavasan** [comment]
@rlrs Thank you for your great analysis here (https://github.com/pytorch/torchtitan/pull/279#issuecomment-2104797493). 

Helped us narrow down the issue which basically boiled down to in-place loading of checkpoint of DCP. StatefulDataLoader doesn't currently return no state if dataloader iterator is not created while DCP expects the module it let it know what the keys the module is expecting. 

In order to get around this, I serialized the state of the dataloader and in this case there is only one key to load that is communicated by the DataLoaderWrapper to DCP - "<rank_id>".

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py:145]: May I ask when is `skip()` not supported? It's kinda neat if we could use the same API to skip...

File: .github/workflows/unit_test_4gpu.yaml
Comment (tianyu-l) [inline .github/workflows/unit_test_4gpu.yaml]: Do we need this?

File: .github/workflows/unit_test_cpu.yaml
Comment (tianyu-l) [inline .github/workflows/unit_test_cpu.yaml]: ditto

File: torchtitan/checkpoint.py
Comment (tianyu-l) [inline torchtitan/checkpoint.py]: wow is this made possible by recent change from StatefulDataloader? What nightly version can we use to run it?

File: test/datasets/test_dataset_checkpoint.py
Comment (tianyu-l) [inline test/datasets/test_dataset_checkpoint.py]: This test logic looks great! IIUC this is not in CI yet and needs to be run manually?

Also I think we need some sort of integration test as well -- basically save checkpoint at step T, and resume from checkpoint and assert we are loading normally at step T+1. In fact we are working on setting up examples to make it easier to add such a test. Let me check when can it land.

**Review by tianyu-l — COMMENTED** [review]
>  I think we can store the dataloader state just for the dp worker ranks (by using key as the dp_rank_id) and load it back instead of storing it for all global ranks.

This sounds quite good! @fegin would checkpointing behave in this expected way? I.e. if we use the same key for the same TP ranks, but different keys for different DP ranks, would it avoid saving extra copies and load correctly? If that's the case I agree we don't have to use DTensor for now.

File: .github/workflows/unit_test_cpu.yaml
Comment (gokulavasan) [reply .github/workflows/unit_test_cpu.yaml]: The pip install? I wasn't able to run the unit test without it, it says torchtitan package is not found

File: torchtitan/checkpoint.py
Comment (gokulavasan) [reply torchtitan/checkpoint.py]: Sorry what is made possible? This is just deserializing the byte string of dataloader state back to a state_dict and passing it in.

File: torchtitan/datasets/hf_datasets.py
Comment (gokulavasan) [reply torchtitan/datasets/hf_datasets.py:145]: Take a look at this notebook - https://colab.research.google.com/drive/15ZeB5yMBey00MJ9MOeK24R1xMn2Hp2AU#scrollTo=6weCenMx9ntp. Basically c4 dataset after split doesn't allow skip. This looks like a bug that we can post in HF github issues to find out more but for now naively iterating. I can maybe add a link to the github issue in the comment

File: test/datasets/test_dataset_checkpoint.py
Comment (gokulavasan) [reply test/datasets/test_dataset_checkpoint.py]: It is in the CI. In cpu unit test yaml file, I see pytest command that runs at the top level. It should pick up the test. When I ran it locally, here is what I see:

>test/datasets/test_dataset_checkpoint.py::TestDatasetCheckpoint::test_c4_resumption PASSED                                                                                                               [ 14%]
test/test_job_config.py::TestJobConfig::test_command_line_args PASSED                                                                                                                                    [ 28%]
test/test_job_config.py::TestJobConfig::test_job_config_file PASSED                                                                                                                                      [ 42%]
test/test_job_config.py::TestJobConfig::test_job_file_does_not_exist PASSED                                                                                                                              [ 57%]
test/test_job_config.py::TestJobConfig::test_empty_config_file PASSED                                                                                                                                    [ 71%]
test/test_job_config.py::TestJobConfig::test_job_config_file_cmd_overrides PASSED                                                                                                                        [ 85%]
test/test_job_config.py::TestJobConfig::test_print_help PASSED                                                                                                                                           [100%]

---------- coverage: platform linux, python 3.10.14-final-0 ----------
Coverage XML written to file coverage.xml

Comment (gokulavasan) [reply test/datasets/test_dataset_checkpoint.py]: > Also I think we need some sort of integration test as well -- basically save checkpoint at step T, and resume from checkpoint and assert we are loading normally at step T+1. 

This would be nice but it might be outside the scope of this PR. The reason being that it this also requires setting up the run to be deterministic etc.

>  In fact we are working on setting up examples to make it easier to add such a test. Let me check when can it land.

Once this is available, we can add an integration test using that

File: .github/workflows/unit_test_4gpu.yaml
Comment (gokulavasan) [reply .github/workflows/unit_test_4gpu.yaml]: Answered below though I am not sure what tests are run for the 4gpu case

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [reply torchtitan/datasets/hf_datasets.py:145]: As discussed offline, I can refactor it after this PR lands, because we'd like to use `skip()` for IterableDataset as well.

File: .github/workflows/unit_test_4gpu.yaml
Comment (tianyu-l) [reply .github/workflows/unit_test_4gpu.yaml]: ditto: Let's remove this `pip install -e ".[dev]"` if not necessary

File: .github/workflows/unit_test_cpu.yaml
Comment (tianyu-l) [reply .github/workflows/unit_test_cpu.yaml]: I just tried: it will work if you have `__init__.py` in the `test/datasets` subfolder. Alternatively you can just put the .py test file in `test/`

File: .github/workflows/unit_test_4gpu.yaml
Comment (tianyu-l) [inline .github/workflows/unit_test_4gpu.yaml:34]: I think we need this in `.github/workflows/integration_test_periodic.yaml` as well.

Let's create an issue, tracking that we need to put torchdata in `requirements.txt` and `pyproject.toml` after the needed changes ship in an official release.

File: torchtitan/datasets/hf_datasets.py
Comment (tianyu-l) [inline torchtitan/datasets/hf_datasets.py]: Let's use `DPAwareDataloader`, because in the [PEP 8 guide](https://peps.python.org/pep-0008/)
> Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.

**Review by tianyu-l — APPROVED** [review]
Looks awesome! Thanks for the beautiful work!

Please address inline comments before merging.

File: .github/workflows/unit_test_4gpu.yaml
Comment (gokulavasan) [reply .github/workflows/unit_test_4gpu.yaml:34]: @tianyu-l Should pyproject.toml datasets dependency also enforce >= 2.19.0 version requirement?

Comment (tianyu-l) [reply .github/workflows/unit_test_4gpu.yaml:34]: ah yes I think so

Comment (gokulavasan) [reply .github/workflows/unit_test_4gpu.yaml:34]: Created https://github.com/pytorch/torchtitan/issues/351 to remove torchdata nightly pip install

**Comment by lhoestq** [comment]
Hi ! I'm Quentin from HF :)
FYI we just added state_dict() and load_state_dict() in `datasets.IterableDataset`, which can resume iteration faster than just skipping samples !

**Pull request merged**
